Keywords/webUtils/WebTableDataComparison.groovy

```diff
@@ -0,0 +1,209 @@
+			}
+			for (String keyName : entriesOnlyOnRight.keySet()) {
+				String actValue = " ** NOT FOUND ** ";
+				String expValue = entriesOnlyOnRight.get(keyName);
+				KeywordUtil.markFailed("Failed " + recordKeyName + " :: " + recordKeyValue + " " + actValue + " " + expValue)
+			}
+			for (String keyName : entriesDiffering.keySet()) {
+				String actValue = entriesDiffering.get(keyName).leftValue();
+				String expValue = entriesDiffering.get(keyName).rightValue();
+				KeywordUtil.markFailed("Failed " + recordKeyName + " :: " + recordKeyValue + " " + actValue + " " + expValue)
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+			KeywordUtil.markFailed("Exception in reportMapDifference, unable to find differences in WebTables")
+		}
+	}
+
+	/**
+	 * Get the WebTable as Map
+	 * @param WebElement of WebTable
+	 * @param PrimaryKey of WebTable
+	 */
+	public static Map<String, Map<String, String>> getWebTableAsMap(WebElement table, String keyName) {
+		Map<String, Map<String, String>> webTableRecords = new LinkedHashMap();
+		int recordCounter = 0;
+		try {
+			//Get the WebTable into a List of String
+			List<String> webTableData = readWebTableData(table);
+			String colNames = webTableData.get(0);
+			String[] arrColName = colNames.split(",", -1);
+			int columnCount = arrColName.length;
+			for (String colValues : webTableData) {
+				String[] arrColValue = colValues.split(",", -1);
+				Map<String, String> fileRecord = new LinkedHashMap<>();
+				String mapKey = "";
+				for (int i = 0; i < columnCount; i++) {
+					String columnName = arrColName[i];
+					if (columnName.equalsIgnoreCase(keyName))
+						mapKey = arrColValue[i];
+					fileRecord.put(columnName, arrColValue[i]);
+				}
+				if (keyName.equalsIgnoreCase("DEFAULT") || mapKey.equalsIgnoreCase("")) {
+					webTableRecords.put(String.valueOf(recordCounter), fileRecord);
+					recordCounter++;
+				} else
+					webTableRecords.put(mapKey, fileRecord);
+			}
+		} catch (Exception e) {
+			e.printStackTrace();
+			KeywordUtil.markFailed("Exception in getWebTableAsMap, unable to get WebTable as Map")
+		}
+		return webTableRecords;
+	}
+
+	/**
+	 * Read the WebTable Data records
+	 * @param WebElement of WebTable
+	 */
+	public static List<String> readWebTableData(WebElement tableElement) {
+		List<String> tableData = new ArrayList<>();
+		try {
+			// Get all rows
+			List<WebElement> rows = tableElement.findElements(By.tagName("tr"));
+			for (WebElement row : rows) {
+				// Get all cells in the current row
+				String rowData = null;
+				List<WebElement> headers = row.findElements(By.tagName("th"));
+				List<WebElement> cells = row.findElements(By.tagName("td"));
+				int count = 0;
+				for (WebElement header : headers) {
+					// Extract data from each header and add it to the list
+					if (count == 0) {
+						rowData = header.getText();
+					} else {
+						rowData = rowData + "," + header.getText();
+					}
+					count++;
+				}
```

Comment:
Why are we fetching header again and again before reading every row? It should be out of for loop

Reply:
Updated the code as suggested